app/uk/gov/hmrc/gform/sharedmodel/VariadicFormData.scala

```diff
@@ -213,8 +216,103 @@ object VariadicFormData {
     // properly classify data to VariadicValues
     val multiValueIds: Set[BaseComponentId] = formModel.allMultiSelectionIds.map(_.baseComponentId)
 
+    val dataGroupBy = data.groupBy(_._1.baseComponentId)
+
+    val allGroupFormComponentsMap = formModel.allFormComponents.collect {
+      case fc @ IsGroup(group) => fc.baseComponentId -> (fc.baseComponentId +: group.fields.map(_.baseComponentId))
+    }.toMap
+
+    def isPureExists(ids: List[ModelComponentId]): Boolean =
+      ids.exists { r =>
+        r match {
+          case Pure(iPure(BaseComponentId(_))) => true
+          case _                               => false
+        }
+      }
+
+    def isAtomicExists(ids: List[ModelComponentId]): Boolean =
+      ids.exists { r =>
+        r match {
+          case Atomic(iPure(BaseComponentId(_)), _) => true
+          case _                                    => false
+        }
+      }
```

Comment:
this could be simplified 

`def isAtomicExists(ids: List[ModelComponentId]): Boolean =  ids.foldLeft(false)((acc, id) => acc && id.fold(_ => false)(a => a.fold(_ => true)(_ => false)))`